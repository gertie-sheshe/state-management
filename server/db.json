{
  "users": [
    {
      "id": 1,
      "name": "Leanne Graham",
      "username": "Bret",
      "email": "Sincere@april.biz"
    },
    {
      "id": 2,
      "name": "Ervin Howell",
      "username": "Antonette",
      "email": "Shanna@melissa.tv"
    },
    {
      "id": 3,
      "name": "Clementine Bauch",
      "username": "Samantha",
      "email": "Nathan@yesenia.net"
    },
    {
      "id": 4,
      "name": "Patricia Lebsack",
      "username": "Karianne",
      "email": "Julianne.OConner@kory.org"
      }
  ],
  "todos": [
    {
      "userId": 1,
      "id": 1,
      "title": "delectus aut autem",
      "completed": false
    },
    {
      "userId": 2,
      "id": 2,
      "title": "quis ut nam facilis et officia qui",
      "completed": false
    },
    {
      "userId": 3,
      "id": 3,
      "title": "fugiat veniam minus",
      "completed": false
    },
    {
      "userId": 4,
      "id": 4,
      "title": "et porro tempora",
      "completed": true
    },
    {
      "userId": 1,
      "id": 5,
      "title": "laboriosam mollitia et enim quasi adipisci quia provident illum",
      "completed": false
    },
    {
      "userId": 2,
      "id": 6,
      "title": "qui ullam ratione quibusdam voluptatem quia omnis",
      "completed": false
    },
    {
      "userId": 3,
      "id": 7,
      "title": "illo expedita consequatur quia in",
      "completed": false
    },
    {
      "userId": 4,
      "id": 8,
      "title": "quo adipisci enim quam ut ab",
      "completed": true
    },
    {
      "userId": 1,
      "id": 9,
      "title": "molestiae perspiciatis ipsa",
      "completed": false
    },
    {
      "userId": 2,
      "id": 10,
      "title": "illo est ratione doloremque quia maiores aut",
      "completed": true
    },
    {
      "userId": 3,
      "id": 11,
      "title": "vero rerum temporibus dolor",
      "completed": true
    },
    {
      "userId": 4,
      "id": 12,
      "title": "ipsa repellendus fugit nisi",
      "completed": true
    },
    {
      "userId": 1,
      "id": 13,
      "title": "et doloremque nulla",
      "completed": false
    }
  ]
}