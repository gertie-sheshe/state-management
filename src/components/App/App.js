import React, { useEffect } from "react";
import Users from "../Users";
import TodoList from "../TodosList";
import TodoForm from "../TodoForm";
import { usersSelectors, fetchUsers } from "../../features/usersSlice";
import { useDispatch, useSelector } from "react-redux";
import "./App.css";

function App() {
  const dispatch = useDispatch();
  const users = useSelector(usersSelectors.selectAll);

  useEffect(() => {
    dispatch(fetchUsers());
  }, []);

  return (
    <div className="App">
      <h1>Todos</h1>
      <main>
        <Users users={users} />
        <TodoList />
        <TodoForm />
      </main>
    </div>
  );
}

export default App;
