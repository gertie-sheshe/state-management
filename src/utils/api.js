const API_URL = "http://localhost:3000";
const HEADERS = {
  Accept: "application/json",
  "Content-Type": "application/json",
};

export const apiRequest = async (url, options) => {
  const result = await fetch(url, options);
  const data = await result.json();

  return data;
};

export const getUsers = () => {
  return apiRequest(`${API_URL}/users`);
};

export const getTodos = () => {
  return apiRequest(`${API_URL}/todos`);
};

export const getUserTodos = (userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`);
};

export const updateTodo = (todoId, data) => {
  return apiRequest(`${API_URL}/todos/${todoId}`, {
    method: "PUT",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};

export const createTodo = (data, userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`, {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};
