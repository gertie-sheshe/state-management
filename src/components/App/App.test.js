import { render } from "../../tests/utils";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { worker } from "../../tests/mocks/server";
import { queryClient } from "../../tests/utils.js";
import { fetchUsersError, fetchUsersNoTodos } from "../../tests/mocks/handlers";

import App from "./App";

describe("App", () => {
  beforeAll(() => worker.listen({ onUnhandledRequest: "warn" }));
  afterEach(() => {
    worker.resetHandlers();
    queryClient.clear();
  });
  afterAll(() => {
    worker.close();
  });
  test("Matches snapshot", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Shows "Select User" option', async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
  });

  test("Shows users", async () => {
    render(<App />);
    expect(await screen.findByText("Leanne Graham")).toBeInTheDocument();
    expect(await screen.findByText("Ervin Howell")).toBeInTheDocument();
    expect(await screen.findByText("Clementine Bauch")).toBeInTheDocument();
    expect(await screen.findByText("Patricia Lebsack")).toBeInTheDocument();
  });

  test('Should correctly set default option to "Select User"', async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    expect(
      screen.getByRole("option", { name: "Select User", selected: true })
    ).toBeInTheDocument();
  });

  it("should display the correct number of options", async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    expect(screen.getAllByRole("option").length).toBe(5);
  });

  test('Should correctly set selected option to "Ervin Howell"', async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    userEvent.selectOptions(screen.getByRole("combobox"), "2");

    await waitFor(() => {
      expect(
        screen.getByRole("option", { name: "Ervin Howell", selected: true })
      ).toBeInTheDocument();
    });
  });

  test("Should enable Add Todo button when a user is selected", async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    userEvent.selectOptions(screen.getByRole("combobox"), "2");

    await waitFor(() => {
      expect(screen.getByRole("button", { name: "Add" }).disabled).toBe(false);
    });
  });

  test("Should disable Add Todo button when a user is not selected", async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByRole("button", { name: "Add" }).disabled).toBe(true);
    });
  });

  test("Should show error message when fetch users fails", async () => {
    worker.use(fetchUsersError);
    render(<App />);
    expect(await screen.findByText("Error:")).toBeInTheDocument();
  });

  test("Should show list of todos when a user is selected", async () => {
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    userEvent.selectOptions(screen.getByRole("combobox"), "2");

    await waitFor(() => {
      expect(screen.getByText("delectus aut autem")).toBeInTheDocument();
    });

    expect(
      screen.getByText(
        "laboriosam mollitia et enim quasi adipisci quia provident illum"
      )
    ).toBeInTheDocument();
    expect(screen.getByText("molestiae perspiciatis ipsa")).toBeInTheDocument();
    expect(screen.getByText("et doloremque nulla")).toBeInTheDocument();
    expect(screen.getByText("home alone")).toBeInTheDocument();
  });

  test("Should show no todos when a user is selected and has no todos", async () => {
    worker.use(fetchUsersNoTodos);
    render(<App />);
    expect(await screen.findByText("Select User")).toBeInTheDocument();
    userEvent.selectOptions(screen.getByRole("combobox"), "4");

    await waitFor(() => {
      expect(screen.getByText("No todos found")).toBeInTheDocument();
    });
  });
});
