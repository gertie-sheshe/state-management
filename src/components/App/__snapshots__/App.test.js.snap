// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1>
      Todos
    </h1>
    <main>
      <p>
        Loading...
      </p>
      <form>
        <ul
          style="list-style-type: none;"
        />
      </form>
      <div
        style="margin-top: 20px;"
      >
        <form>
          <label
            for="todo"
            style="margin-right: 10px;"
          >
            Add Todo:
          </label>
          <input
            disabled=""
            id="todo"
            type="text"
            value=""
          />
          <button
            disabled=""
            style="margin-left: 5px;"
            type="submit"
          >
            Add
          </button>
        </form>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
